driver-core/src/main/java/com/datastax/driver/core/Message.java

```diff
@@ -170,21 +171,25 @@ public UUID getTracingId() {
         }
     }
 
-    public static class ProtocolDecoder extends OneToOneDecoder {
+    @ChannelHandler.Sharable
+    public static class ProtocolDecoder extends MessageToMessageDecoder<Frame> {
 
-        public Object decode(ChannelHandlerContext ctx, Channel channel, Object msg) {
-            assert msg instanceof Frame : "Expecting frame, got " + msg;
-
-            Frame frame = (Frame)msg;
+        @Override
+        protected void decode(ChannelHandlerContext ctx, Frame frame, List<Object> out) throws Exception {
             boolean isTracing = frame.header.flags.contains(Frame.Header.Flag.TRACING);
             UUID tracingId = isTracing ? CBUtil.readUUID(frame.body) : null;
 
             Response response = Response.Type.fromOpcode(frame.header.opcode).decoder(frame.header.version).decode(frame.body);
-            return response.setTracingId(tracingId).setStreamId(frame.header.streamId);
+            ReferenceCountUtil.release(frame.body);
```

Comment:
`frame.body` comes from `Frame.Decoder.decode`. It's the buffer that was returned by Netty because of the way `Frame.create` works (it consumes the beginning of the buffer to decode the header, and then keeps the buffer as body).
Since it was created by Netty we have to deallocate it. But we can't do it in `Frame.Decoder.decode` because we hold onto it until line 182 here, where it gets decoded. Once that's done we won't need it anymore.
I put a comment in the frame decoder, maybe we should add one here too.